Document `cml runner` Bitbucket support (#243)
iterative/cml.dev: content/docs/self-hosted-runners.md

# Self-hosted (On-premise or Cloud) Runners

> GitHub Actions, GitLab CI/CD and Bitbucket Pipelines workflows are run on
> GitHub-, GitLab- and Bitbucket- hosted runners by default. However, there are

Reviewer: ```suggestion
GitHub Actions, GitLab CI/CD, and Bitbucket Pipelines workflows are executed on
"native" runners (hosted by GitHub/GitLab/Bitbucket respectively) by default.
However, there are
```

many great reasons to use your own runners: to take advantage of GPUs,
orchestrate your team's shared computing resources, or train in the cloud.

> ☝️ **Tip!** Check out the official documentation from

Reviewer: ```suggestion
<admon type="tip">

Check out the official documentation from
```

[GitHub](https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners),
[GitLab](https://docs.gitlab.com/runner) and
[Bitbucket](https://support.atlassian.com/bitbucket-cloud/docs/runners) for more

> information on self-hosted runners.

Reviewer: ```suggestion
information on self-hosted runners.

</admon>
```

      - uses: actions/checkout@v2
      - name: Train model
        env:

>           REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}

Author: There is a reason for this change: the `PERSONAL_ACCESS_TOKEN` name is arbitrary, and `cml` only reads the `REPO_TOKEN` environment variable; forges other than GitHub don't have explicit secrets, so it's better to unify the names.

  script:
    - pip install -r requirements.txt
    - python train.py
    # Create CML report
    - cat metrics.txt >> report.md
    - cml publish plot.png --md --title="Confusion Matrix" >> report.md
    - cml send-comment report.md
```

</tab>
<tab title="Bitbucket">

```yaml
pipelines:
  default:
    - step:
        image: iterativeai/cml:0-dvc2-base1
        script:
          - |
            cml runner \
                --cloud=aws \
                --cloud-region=us-west \
                --cloud-type=p2.xlarge \
                --cloud-spot \

>                 --labels=cml-gpu
>     - step:
>         runs-on: [self.hosted, cml-gpu]

Reviewer: when in Rome?

```suggestion
                --labels=cml.gpu
    - step:
        runs-on: [self.hosted, cml.gpu]
```

>         image: iterativeai/cml:0-dvc2-base1-gpu

Reviewer: do GPUs actually work on BB?

Author: No, please see the callout on https://github.com/iterative/cml.dev/pull/243#issue-1246387011

Author: Should we add a warning and/or remove mentions to GPU until https://github.com/iterative/cml/issues/1015 is resolved?

Reviewer: I'd say:

```diff
- --labels=cml.gpu
- --cloud-type=p2.xlarge
- image: iterativeai/cml:0-dvc2-base1-gpu
+ # gpu not yet supported, see https://github.com/iterative/cml/issues/1015
```

Author: https://github.com/iterative/cml.dev/pull/243#discussion_r882186255

  script:
    - pip install -r requirements.txt
    - python train.py
    # Create CML report
    - cat metrics.txt >> report.md
    - cml publish plot.png --md --title="Confusion Matrix" >> report.md
    - cml send-comment report.md
```

</tab>
<tab title="Bitbucket">

```yaml
pipelines:
  default:
    - step:
        image: iterativeai/cml:0-dvc2-base1
        script:
          - |
            cml runner \
                --cloud=aws \

>                 --cloud-region=us-west \

Reviewer: ```suggestion
                --cloud-region=us-west \
                --cloud-type=m5.2xlarge \
```

>                 --cloud-spot

Reviewer: ```suggestion
                --cloud-spot \
                --labels=cml
```

    - step:

>         runs-on: [self.hosted]

Reviewer: ```suggestion
        runs-on: [self.hosted, cml]
```

  script:
    - pip install -r requirements.txt
    - python train.py
    # Create CML report
    - cat metrics.txt >> report.md
    - cml publish plot.png --md --title="Confusion Matrix" >> report.md
    - cml send-comment report.md
```

</tab>
<tab title="Bitbucket">

```yaml
pipelines:
  default:
    - step:
        image: iterativeai/cml:0-dvc2-base1
        script:
          - |
            cml runner \
                --cloud=aws \
                --cloud-region=us-west \

>                 --cloud-type=p2.xlarge \
>                 --cloud-spot \
>                 --labels=cml.gpu
>     - step:
>         runs-on: [self.hosted, cml.gpu]
>         image: iterativeai/cml:0-dvc2-base1-gpu

Author: ```suggestion
                --cloud-spot
    - step:
        runs-on: [self.hosted]
        image: iterativeai/cml:0-dvc2-base1
```

Sensitive values like cloud and repository credentials can be provided through
environment variables with the aid of GitHub
[secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository),
GitLab
[masked variables](https://docs.gitlab.com/ee/ci/variables/#add-a-cicd-variable-to-a-project)
(or [external secrets](https://docs.gitlab.com/ee/ci/secrets) for added
security), or Bitbucket
[secured user-defined variables](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/#User-defined-variables).

> ⚠️ You will need to create a [personal access token](#personal-access-token)
> with enough permissions to register self-hosted runners. In the example workflow
> above, this token is stored as `REPO_TOKEN`.

Reviewer: ```suggestion
<admon type="warn">

You will need to create a [personal access token (PAT)](#personal-access-token)
with enough permissions to register self-hosted runners. In the example workflow
above, this token is stored as `REPO_TOKEN`.

</admon>
```
